arangod/Aql/SingleRowFetcher.h

```diff
@@ -88,10 +88,37 @@ class SingleRowFetcher {
   // TODO enable_if<passBlocksThrough>
   std::pair<ExecutionState, std::shared_ptr<AqlItemBlockShell>> fetchBlockForPassthrough(size_t atMost);
 
-  std::pair<ExecutionState, size_t> preFetchNumberOfRows() {
-    // This is not implemented for this fetcher
-    TRI_ASSERT(false);
-    THROW_ARANGO_EXCEPTION(TRI_ERROR_INTERNAL);
+  std::pair<ExecutionState, size_t> preFetchNumberOfRows(size_t atMost) {
+    if (_upstreamState == ExecutionState::WAITING ||
+        (_upstreamState == ExecutionState::HASMORE &&
+         (_currentBlock == nullptr /* nothing fetched */ ||
+          _currentBlock->block().size() >= _rowIndex /* fully consumed */))) {
```

Comment:
Can this be simplified, e.g. to `if (!indexIsValid())` or `if (!indexIsValid() && _upstreamState != ExecutionState::DONE)`? We don't fetch a new block unless the index is invalid, thus `_upstreamState == ExecutionState::WAITING` should imply `!indexIsValid()`, or am I missing something?

Reply:
Maybe it would also help to simplify the conditions if this and the following cases are reordered.

Reply:
We cannot reorder the conditions, the first conditional block can activate the second one.
e.g fetch a new block, which is the last from above.